scylla/src/transport/connection_pool.rs

```diff
@@ -907,45 +908,66 @@ impl PoolRefiller {
     // to the shard using the port.
     fn start_opening_connection(&self, shard: Option<Shard>) {
-        // temporary in this commit
-        let (address, regular_port) = {
-            let addr = self.node_addr.inner();
-            (addr.ip(), addr.port())
-        };
-
         let cfg = self.pool_config.connection_config.clone();
-        let fut = match (self.sharder.clone(), self.shard_aware_port, shard) {
-            (Some(sharder), Some(port), Some(shard)) => {
-                let shard_aware_address = (address, port).into();
-                async move {
-                    let result = open_connection_to_shard_aware_port(
-                        shard_aware_address,
-                        shard,
-                        sharder.clone(),
-                        &cfg,
-                    )
-                    .await;
-                    Ok(OpenedConnectionEvent {
-                        result,
-                        requested_shard: Some(shard),
-                        keyspace_name: None,
-                    })
+
+        let node_addr = self.node_addr;
+        let address_translator = self.address_translator.clone();
+        let addr = async move {
+            match node_addr {
+                NodeAddr::RealPeer(addr) => {
+                    // In this case, addr is subject to AddressTranslator.
+                    if let Some(translator) = address_translator {
+                        let res = translator
+                            .translate_address(&UntranslatedPeer {
+                                untranslated_address: addr,
+                            })
+                            .await;
+                        if let Err(ref err) = res {
+                            error!("Address translation failed for addr {}: {}", addr, err);
+                        }
+                        res
+                    } else {
+                        Ok(addr)
+                    }
                 }
-                .boxed()
-            }
-            _ => {
-                let non_shard_aware_address = (address, regular_port).into();
-                async move {
-                    let result =
-                        connection::open_connection(non_shard_aware_address, None, cfg).await;
-                    Ok(OpenedConnectionEvent {
-                        result,
-                        requested_shard: None,
-                        keyspace_name: None,
-                    })
+                NodeAddr::InitiallyKnown(addr) => {
+                    // Known nodes, a.k.a. contact points, are user-provied addresses that do not require translation.
+                    Ok(addr)
                 }
-                .boxed()
+                NodeAddr::FixedLocal(addr) => {
+                    // In this case, addr is considered to be translated, as it is the control connection's address.
+                    Ok(addr)
+                }
+            }
+        };
```

Comment:
I think this code could be pushed down to an even lower layer - maybe put the address translator to the connection config and then use it in `connection::open_conection`?

Moreover, you aren't considering the shard-aware port in this code, but you could.